app/src/main/java/com/kabouzeid/gramophone/dialogs/DeleteSongsDialog.java

```diff
@@ -54,14 +69,162 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
                 .content(content)
                 .positiveText(R.string.delete_action)
                 .negativeText(android.R.string.cancel)
+                .autoDismiss(false)
                 .onPositive(new MaterialDialog.SingleButtonCallback() {
                     @Override
                     public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
-                        if (getActivity() == null)
-                            return;
-                        MusicUtil.deleteTracks(getActivity(), songs);
+                        songsToRemove = songs;
+                        new DeleteSongsAsyncTask(DeleteSongsDialog.this).execute(new DeleteSongsAsyncTask.LoadingInfo(songs, null));
+                    }
+                })
+                .onNegative(new MaterialDialog.SingleButtonCallback() {
+                    @Override
+                    public void onClick(@NonNull MaterialDialog materialDialog, @NonNull DialogAction dialogAction) {
+                        dismiss();
                     }
                 })
                 .build();
     }
+
+    private void deleteSongs(List<Song> songs, List<Uri> safUris) {
+        MusicUtil.deleteTracks(getActivity(), songs, safUris);
+    }
+
+    @TargetApi(Build.VERSION_CODES.KITKAT)
+    private void deleteSongsKitkat() {
+        if (songsToRemove.size() < 1) {
+            dismiss();
+            return;
+        }
+
+        currentSong = songsToRemove.remove(0);
+
+        if (!SAFUtil.isSAFRequired(currentSong)) {
+            deleteSongs(Collections.singletonList(currentSong), null);
+            deleteSongsKitkat();
+        } else {
+            Toast.makeText(getActivity(), String.format(getString(R.string.saf_pick_file), currentSong.data), Toast.LENGTH_LONG).show();
+            SAFUtil.openFilePicker(this);
+        }
+    }
+
+    @Override
+    public void onActivityResult(int requestCode, int resultCode, Intent intent) {
+        super.onActivityResult(requestCode, resultCode, intent);
+        switch (requestCode) {
+            case SAFGuideActivity.REQUEST_CODE_SAF_GUIDE:
+                SAFUtil.openTreePicker(this);
+                break;
+
+            case SAFUtil.REQUEST_SAF_PICK_TREE:
+            case SAFUtil.REQUEST_SAF_PICK_FILE:
+                new DeleteSongsAsyncTask(this).execute(new DeleteSongsAsyncTask.LoadingInfo(requestCode, resultCode, intent));
+                break;
+        }
+    }
+
+    private static class DeleteSongsAsyncTask extends DialogAsyncTask<DeleteSongsAsyncTask.LoadingInfo, Integer, Void> {
+        private DeleteSongsDialog dialog;
+        private Activity activity;
```

Comment:
You should use `WeakReference`s for these in case the activity (and potentially dialog?) are killed off before the task is done.